datafusion/core/src/physical_optimizer/projection_pushdown.rs

```diff
@@ -0,0 +1,2181 @@
+            .unwrap_or(false)
+    })
+}
+
+/// Updates a source provider's projected columns according to the given
+/// projection operator's expressions. To use this function safely, one must
+/// ensure that all expressions are `Column` expressions without aliases.
+fn new_projections_for_columns(
+    projection: &ProjectionExec,
+    source: &Option<Vec<usize>>,
+) -> Vec<usize> {
+    projection
+        .expr()
+        .iter()
+        .filter_map(|(expr, _)| {
+            expr.as_any()
+                .downcast_ref::<Column>()
+                .and_then(|expr| source.as_ref().map(|proj| proj[expr.index()]))
+        })
+        .collect()
+}
+
+/// The function operates in two modes:
+///
+/// 1) When `sync_with_child` is `true`:
+///
+///    The function updates the indices of `expr` if the expression resides
+///    in the input plan. For instance, given the expressions `a@1 + b@2`
+///    and `c@0` with the input schema `c@2, a@0, b@1`, the expressions are
+///    updated to `a@0 + b@1` and `c@2`.
+///
+/// 2) When `sync_with_child` is `false`:
+///
+///    The function determines how the expression would be updated if a projection
+///    was placed before the plan associated with the expression. If the expression
+///    cannot be rewritten after the projection, it returns `None`. For example,
+///    given the expressions `c@0`, `a@1` and `b@2`, and the [`ProjectionExec`] with
+///    an output schema of `a, c_new`, then `c@0` becomes `c_new@1`, `a@1` becomes
+///    `a@0`, but `b@2` results in `None` since the projection does not include `b`.
+///
+/// If the expression contains a `PhysicalExpr` variant that this function does
+/// not support, it will return `None`. An error can only be introduced if
+/// `CaseExpr::try_new` returns an error.
+fn update_expr(
+    expr: &Arc<dyn PhysicalExpr>,
+    projected_exprs: &[(Arc<dyn PhysicalExpr>, String)],
+    sync_with_child: bool,
+) -> Result<Option<Arc<dyn PhysicalExpr>>> {
+    let expr_any = expr.as_any();
+    if let Some(column) = expr_any.downcast_ref::<Column>() {
+        if sync_with_child {
+            // Update the index of `column`:
+            Ok(Some(projected_exprs[column.index()].0.clone()))
+        } else {
+            // Determine how to update `column` to accommodate `projected_exprs`:
+            Ok(projected_exprs.iter().enumerate().find_map(
+                |(index, (projected_expr, alias))| {
+                    projected_expr.as_any().downcast_ref::<Column>().and_then(
+                        |projected_column| {
+                            column
+                                .name()
+                                .eq(projected_column.name())
+                                .then(|| Arc::new(Column::new(alias, index)) as _)
+                        },
+                    )
+                },
+            ))
+        }
+    } else if let Some(binary) = expr_any.downcast_ref::<BinaryExpr>() {
```

Comment:
This pattern is basically recursing through the PhysicalExpr tree manually, and only covers some subset of of the nodes.

I tried rewriting it using TreeNode, which is both less code and covers all PhysicalExpr types, not just a subset, and the tests still pass. 

I will make a follow on PR with the proposed simplification

Reply:
https://github.com/apache/arrow-datafusion/pull/8109#